infra/modules/service/workflow_orchestrator_role.tf

```
@@ -0,0 +1,110 @@
+#--------------------------------
+# Scheduler Workflow Manager Role
+#--------------------------------
+# This role and policy are used by the Step Functions state machine that manages the scheduled jobs workflow.
+
+resource "aws_iam_role" "workflow_orchestrator" {
+  name                = "${var.service_name}-workflow-orchestrator"
+  managed_policy_arns = [aws_iam_policy.workflow_orchestrator.arn]
+  assume_role_policy  = data.aws_iam_policy_document.workflow_orchestrator_assume_role.json
+}
+
+data "aws_iam_policy_document" "workflow_orchestrator_assume_role" {
+  statement {
+    sid     = "ECSTasksAssumeRole"
+    actions = ["sts:AssumeRole"]
+    principals {
+      type        = "Service"
+      identifiers = ["states.amazonaws.com"]
+    }
+    condition {
+      test     = "ArnLike"
+      variable = "aws:SourceArn"
+      values   = ["arn:aws:states:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:stateMachine:*"]
```

we actually have the state machine ARNs available so instead of "*" let's use those:
```suggestion
      values   = [aws_sfn_state_machine.scheduled_jobs[*].arn]
```

---

The state machine itself uses this role, so unfortunately its a circular dependency.

```
terraform apply -var "environment_name=dev"
╷
│ Error: Cycle: module.service.data.aws_iam_policy_document.workflow_orchestrator_assume_role, module.service.aws_iam_role.workflow_orchestrator, module.service.aws_sfn_state_machine.scheduled_jobs
```

---

Oh ok can we manually construct the ARN then:

```suggestion
      values   = [for job_name in keys(var.scheduled_jobs) "arn:aws:states:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:stateMachine:${var.service_name}-${job_name}"]
```

---

Unfortunately I can't get any get this to work. The error message I'm getting is:

```
The principal states.amazonaws.com is not authorized to assume the provided role. (role: arn:aws:iam::{REDACTED account id}:role/p-130-app-dev-workflow-orchestrator)
```

Which doesn't make sense to me because the arn is an exact match 🤷🏼

---

Alright no big deal we can always refine in the future. Thanks!